lib/wiki_api.rb

```diff
@@ -97,46 +101,36 @@ def mediawiki(action, query)
     tries ||= 3
     @mediawiki = api_client
     @mediawiki.send(action, query)
-  rescue MediawikiApi::ApiError => e
-    log_error e, action, query
   rescue StandardError => e
     tries -= 1
-    log_error e, action, query
-    handle_non_api_error(e)
-    retry if tries >= 0
-    Raven.capture_exception e, level: 'warning'
-    return nil # Do not return a Raven object
+    # Raise unknown errors.
+    # Continue for typical errors so that the request can be retried, but wait
+    # a short bit in the case of 429 — too many request — errors.
+    sleep 1 if too_many_requests?(e)
+    retry unless tries.zero?
+    invoke_error_handling_tasks(e, action, query)
+    raise e unless TYPICAL_ERRORS.include?(e.class)
+    return nil
   end
 
   def api_client
     MediawikiApi::Client.new @api_url
   end
 
-  def log_error(e, action, query)
-    Rails.logger.info "Caught #{e}"
-    Raven.capture_exception e, level: 'warning',
-                               extra: { action: action,
-                                        query: query,
-                                        api_url: @api_url }
-    return nil # Do not return a Raven object
-  end
-
-  # Raise unknown errors.
-  # Continue for typical errors so that the request can be retried, but wait
-  # a short bit in the case of 429 — too many request — errors.
-  def handle_non_api_error(e)
-    raise e unless typical_errors.include?(e.class)
-    sleep 1 if too_many_requests?(e)
-  end
-
   def too_many_requests?(e)
     return false unless e.class == MediawikiApi::HttpError
     e.status == 429
   end
 
-  def typical_errors
-    [Faraday::TimeoutError,
-     Faraday::ConnectionFailed,
-     MediawikiApi::HttpError]
+  TYPICAL_ERRORS = [Faraday::TimeoutError,
+                    Faraday::ConnectionFailed,
+                    MediawikiApi::HttpError,
+                    MediawikiApi::ApiError].freeze
+
+  def invoke_error_handling_tasks(error, action, query)
+    Rails.logger.info "Caught #{error}"
```

Comment:
This type of logging should definitely be in the error handling module and not the API class.

Reply:
Done! Earlier only WikiApi was logging the error, so I thought it might be specific to it.